syntax = "proto3";

package xla.poplarplugin;

import "tensorflow/compiler/xla/xla_data.proto";
import "tensorflow/compiler/plugin/poplar/driver/option_flag.proto";
import "tensorflow/compiler/plugin/poplar/driver/poplar_feed_config.proto";

// Configuration for one of the infeed/outfeed streams
message FeedConfig {
  string stream_prefix = 1;
  PoplarFeedConfig config = 2;
  ShapeProto shape = 3;
}

message SendRecvConfig {
  string stream_handle = 1;
  string rendezvous_key = 2;
  ShapeProto shape = 3;
}

message PoplarExecutableProto {

  // The serialized Engine
  string engine = 1;

  // The number of replicas
  int32 replication_factor = 2;

  // Names of the infeed instructions
  repeated FeedConfig infeeds = 3;

  // Names of the outfeed instructions
  repeated FeedConfig outfeeds = 4;

  // The OptionFlags which were used during engine compilation
  repeated PoplarOptionFlag option_flags = 5;

  repeated SendRecvConfig sends = 6;
  repeated SendRecvConfig recvs = 7;
};
